File: Python/ilinkai_weixin/auth/account_store.py
"""账户存储服务"""
import json
import logging
import os
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional, List

logger = logging.getLogger(__name__)

DEFAULT_BASE_URL = "https://ilinkai.weixin.qq.com"
DEFAULT_CDN_BASE_URL = "https://novac2c.cdn.weixin.qq.com/c2c"


@dataclass
class WeixinAccountData:
    token: Optional[str] = None
    saved_at: Optional[str] = None
    base_url: Optional[str] = None
    user_id: Optional[str] = None


@dataclass
class ResolvedWeixinAccount:
    account_id: str = ""
    base_url: str = DEFAULT_BASE_URL
    cdn_base_url: str = DEFAULT_CDN_BASE_URL
    token: Optional[str] = None
    enabled: bool = True
    configured: bool = False
    name: Optional[str] = None


class AccountStore:
    """账户存储服务 - 管理微信账户的持久化存储"""

    def __init__(self, state_dir: Optional[str] = None):
        self._state_dir = state_dir or self._get_default_state_dir()

    @staticmethod
    def _get_default_state_dir() -> str:
        env_path = os.environ.get('ILINKAI_STATE_DIR')
        if env_path:
            return env_path
        return os.path.join(os.path.expanduser('~'), '.ilinkai')

    def _get_weixin_state_dir(self) -> str:
        return os.path.join(self._state_dir, 'weixin')

    def _get_account_index_path(self) -> str:
        return os.path.join(self._get_weixin_state_dir(), 'accounts.json')

    def _get_accounts_dir(self) -> str:
        return os.path.join(self._get_weixin_state_dir(), 'accounts')

    def _get_account_file_path(self, account_id: str) -> str:
        return os.path.join(self._get_accounts_dir(), f'{self._sanitize_account_id(account_id)}.json')

    @staticmethod
    def _sanitize_account_id(account_id: str) -> str:
        safe = account_id.strip().lower()
        for c in ['/', '\\', ':', '*', '?', '"', '<', '>', '|']:
            safe = safe.replace(c, '_')
        return safe.replace('..', '_')

    @staticmethod
    def normalize_account_id(raw_id: str) -> str:
        trimmed = raw_id.strip().lower()
        if trimmed.endswith('@im.bot'):
            return trimmed.replace('@im.bot', '-im-bot')
        if trimmed.endswith('@im.wechat'):
            return trimmed.replace('@im.wechat', '-im-wechat')
        return trimmed

    @staticmethod
    def derive_raw_account_id(normalized_id: str) -> Optional[str]:
        if normalized_id.endswith('-im-bot'):
            return normalized_id[:-7] + '@im.bot'
        if normalized_id.endswith('-im-wechat'):
            return normalized_id[:-10] + '@im.wechat'
        return None

    def list_account_ids(self) -> List[str]:
        file_path = self._get_account_index_path()
        try:
            if not os.path.exists(file_path):
                return []
            with open(file_path, 'r', encoding='utf-8') as f:
                parsed = json.load(f)
            return [id_ for id_ in (parsed or []) if id_ and id_.strip()]
        except Exception:
            return []

    def register_account_id(self, account_id: str):
        os.makedirs(self._get_weixin_state_dir(), exist_ok=True)
        existing = self.list_account_ids()
        if account_id in existing:
            return
        existing.append(account_id)
        with open(self._get_account_index_path(), 'w', encoding='utf-8') as f:
            json.dump(existing, f, indent=2)

    def load_account(self, account_id: str) -> Optional[WeixinAccountData]:
        file_path = self._get_account_file_path(account_id)
        if os.path.exists(file_path):
            return self._read_account_file(file_path)
        raw_id = self.derive_raw_account_id(account_id)
        if raw_id:
            compat_path = self._get_account_file_path(raw_id)
            if os.path.exists(compat_path):
                return self._read_account_file(compat_path)
        return None

    @staticmethod
    def _read_account_file(file_path: str) -> Optional[WeixinAccountData]:
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            return WeixinAccountData(
                token=data.get('token'), saved_at=data.get('saved_at'),
                base_url=data.get('base_url'), user_id=data.get('user_id'),
            )
        except Exception:
            return None

    def save_account(self, account_id: str, data: WeixinAccountData):
        os.makedirs(self._get_accounts_dir(), exist_ok=True)
        existing = self.load_account(account_id) or WeixinAccountData()
        merged = WeixinAccountData(
            token=(data.token.strip() if data.token else None) or existing.token,
            base_url=(data.base_url.strip() if data.base_url else None) or existing.base_url,
            user_id=data.user_id or existing.user_id,
            saved_at=datetime.now(timezone.utc).isoformat(),
        )
        file_path = self._get_account_file_path(account_id)
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump({
                'token': merged.token, 'saved_at': merged.saved_at,
                'base_url': merged.base_url, 'user_id': merged.user_id,
            }, f, indent=2)

    def clear_account(self, account_id: str):
        try:
            file_path = self._get_account_file_path(account_id)
            if os.path.exists(file_path):
                os.remove(file_path)
        except Exception:
            pass

    def resolve_account(self, account_id: str) -> ResolvedWeixinAccount:
        if not account_id or not account_id.strip():
            raise ValueError("accountId is required")
        id_ = self.normalize_account_id(account_id)
        account_data = self.load_account(id_)
        return ResolvedWeixinAccount(
            account_id=id_,
            base_url=(account_data.base_url.strip() if account_data and account_data.base_url else DEFAULT_BASE_URL),
            cdn_base_url=DEFAULT_CDN_BASE_URL,
            token=(account_data.token.strip() if account_data and account_data.token else None),
            enabled=True,
            configured=bool(account_data and account_data.token and account_data.token.strip()),
        )

File: CSharp/ILinkai.Weixin.Sdk/Auth/AccountStore.cs
using System.Security.Cryptography;
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace ILinkai.Weixin.Sdk.Auth;

/// <summary>
/// 微信账户数据
/// </summary>
public class WeixinAccountData
{
    /// <summary>
    /// 认证令牌
    /// </summary>
    public string? Token { get; set; }

    /// <summary>
    /// 保存时间
    /// </summary>
    public DateTime? SavedAt { get; set; }

    /// <summary>
    /// 基础URL
    /// </summary>
    public string? BaseUrl { get; set; }

    /// <summary>
    /// 关联的微信用户ID
    /// </summary>
    public string? UserId { get; set; }
}

/// <summary>
/// 已解析的微信账户信息
/// </summary>
public class ResolvedWeixinAccount
{
    /// <summary>
    /// 账户ID
    /// </summary>
    public string AccountId { get; set; } = string.Empty;

    /// <summary>
    /// 基础URL
    /// </summary>
    public string BaseUrl { get; set; } = WeixinApiClient.DefaultBaseUrl;

    /// <summary>
    /// CDN基础URL
    /// </summary>
    public string CdnBaseUrl { get; set; } = WeixinApiClient.DefaultCdnBaseUrl;

    /// <summary>
    /// 认证令牌
    /// </summary>
    public string? Token { get; set; }

    /// <summary>
    /// 是否启用
    /// </summary>
    public bool Enabled { get; set; } = true;

    /// <summary>
    /// 是否已配置（已获取令牌）
    /// </summary>
    public bool Configured { get; set; }

    /// <summary>
    /// 账户名称
    /// </summary>
    public string? Name { get; set; }
}

/// <summary>
/// 账户存储服务
/// 管理微信账户的持久化存储
/// </summary>
public class AccountStore : IAccountStore
{
    private readonly string _stateDir;
    private readonly ILogger? _logger;
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        PropertyNameCaseInsensitive = true
    };

    /// <summary>
    /// 创建AccountStore实例
    /// </summary>
    /// <param name="stateDir">状态目录路径</param>
    /// <param name="logger">日志记录器</param>
    public AccountStore(string? stateDir = null, ILogger? logger = null)
    {
        _stateDir = stateDir ?? GetDefaultStateDir();
        _logger = logger;
    }

    /// <summary>
    /// 获取默认状态目录
    /// </summary>
    private static string GetDefaultStateDir()
    {
        var envPath = Environment.GetEnvironmentVariable("ILINKAI_STATE_DIR");
        if (!string.IsNullOrWhiteSpace(envPath)) return envPath;

        var homeDir = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        return Path.Combine(homeDir, ".ilinkai");
    }

    /// <summary>
    /// 获取微信状态目录
    /// </summary>
    private string GetWeixinStateDir() => Path.Combine(_stateDir, "weixin");

    /// <summary>
    /// 获取账户索引文件路径
    /// </summary>
    private string GetAccountIndexPath() => Path.Combine(GetWeixinStateDir(), "accounts.json");

    /// <summary>
    /// 获取账户目录
    /// </summary>
    private string GetAccountsDir() => Path.Combine(GetWeixinStateDir(), "accounts");

    /// <summary>
    /// 获取账户文件路径
    /// </summary>
    private string GetAccountFilePath(string accountId) => Path.Combine(GetAccountsDir(), $"{SanitizeAccountId(accountId)}.json");

    /// <summary>
    /// 清理账户ID中的特殊字符
    /// </summary>
    private static string SanitizeAccountId(string accountId)
    {
        var safe = accountId.Trim().ToLowerInvariant();
        foreach (var c in Path.GetInvalidFileNameChars())
        {
            safe = safe.Replace(c, '_');
        }
        return safe.Replace("..", "_");
    }

    /// <summary>
    /// 规范化账户ID
    /// 将原始ID转换为文件系统安全的格式
    /// </summary>
    public static string NormalizeAccountId(string rawId)
    {
        var trimmed = rawId.Trim().ToLowerInvariant();
        if (trimmed.EndsWith("@im.bot"))
        {
            return trimmed.Replace("@im.bot", "-im-bot");
        }
        if (trimmed.EndsWith("@im.wechat"))
        {
            return trimmed.Replace("@im.wechat", "-im-wechat");
        }
        return trimmed;
    }

    /// <summary>
    /// 从规范化ID推导原始ID
    /// </summary>
    public static string? DeriveRawAccountId(string normalizedId)
    {
        if (normalizedId.EndsWith("-im-bot"))
        {
            return $"{normalizedId.Substring(0, normalizedId.Length - 7)}@im.bot";
        }
        if (normalizedId.EndsWith("-im-wechat"))
        {
            return $"{normalizedId.Substring(0, normalizedId.Length - 10)}@im.wechat";
        }
        return null;
    }

    /// <summary>
    /// 列出所有已注册的账户ID
    /// </summary>
    public List<string> ListAccountIds()
    {
        var filePath = GetAccountIndexPath();
        try
        {
            if (!File.Exists(filePath)) return new List<string>();
            var raw = File.ReadAllText(filePath);
            var parsed = JsonSerializer.Deserialize<List<string>>(raw);
            return parsed?.Where(id => !string.IsNullOrWhiteSpace(id)).ToList() ?? new List<string>();
        }
        catch (Exception ex)
        {
            _logger?.LogWarning(ex, "Failed to list account ids from {FilePath}", filePath);
            return new List<string>();
        }
    }

    /// <summary>
    /// 注册账户ID到索引
    /// </summary>
    public void RegisterAccountId(string accountId)
    {
        var dir = GetWeixinStateDir();
        Directory.CreateDirectory(dir);

        var existing = ListAccountIds();
        if (existing.Contains(accountId)) return;

        existing.Add(accountId);
        File.WriteAllText(GetAccountIndexPath(), JsonSerializer.Serialize(existing, JsonOptions));
    }

    /// <summary>
    /// 加载账户数据
    /// </summary>
    public WeixinAccountData? LoadAccount(string accountId)
    {
        var filePath = GetAccountFilePath(accountId);
        if (File.Exists(filePath))
        {
            return ReadAccountFile(filePath);
        }

        var rawId = DeriveRawAccountId(accountId);
        if (rawId != null)
        {
            var compatPath = GetAccountFilePath(rawId);
            if (File.Exists(compatPath))
            {
                return ReadAccountFile(compatPath);
            }
        }

        return null;
    }

    /// <summary>
    /// 读取账户文件
    /// </summary>
    private WeixinAccountData? ReadAccountFile(string filePath)
    {
        try
        {
            var raw = File.ReadAllText(filePath);
            return JsonSerializer.Deserialize<WeixinAccountData>(raw, JsonOptions);
        }
        catch (Exception ex)
        {
            _logger?.LogWarning(ex, "Failed to read account file {FilePath}", filePath);
            return null;
        }
    }

    /// <summary>
    /// 保存账户数据
    /// </summary>
    public void SaveAccount(string accountId, WeixinAccountData data)
    {
        var dir = GetAccountsDir();
        Directory.CreateDirectory(dir);

        var existing = LoadAccount(accountId) ?? new WeixinAccountData();

        var merged = new WeixinAccountData
        {
            Token = data.Token?.Trim() ?? existing.Token,
            BaseUrl = data.BaseUrl?.Trim() ?? existing.BaseUrl,
            UserId = data.UserId ?? existing.UserId,
            SavedAt = DateTime.UtcNow
        };

        var filePath = GetAccountFilePath(accountId);
        File.WriteAllText(filePath, JsonSerializer.Serialize(merged, JsonOptions));

        try
        {
            File.SetAttributes(filePath, FileAttributes.Normal);
        }
        catch
        {
        }
    }

    /// <summary>
    /// 清除账户数据
    /// </summary>
    public void ClearAccount(string accountId)
    {
        try
        {
            var filePath = GetAccountFilePath(accountId);
            if (File.Exists(filePath))
            {
                File.Delete(filePath);
            }

            var rawId = DeriveRawAccountId(accountId);
            if (rawId != null)
            {
                var compatPath = GetAccountFilePath(rawId);
                if (File.Exists(compatPath))
                {
                    File.Delete(compatPath);
                }
            }

            var existing = ListAccountIds();
            if (existing.Remove(accountId))
            {
                File.WriteAllText(GetAccountIndexPath(), JsonSerializer.Serialize(existing, JsonOptions));
            }
        }
        catch (Exception ex)
        {
            _logger?.LogWarning(ex, "Failed to clear account {AccountId}", accountId);
        }
    }

    /// <summary>
    /// 解析账户信息
    /// </summary>
    public ResolvedWeixinAccount ResolveAccount(string accountId)
    {
        if (string.IsNullOrWhiteSpace(accountId))
        {
            throw new ArgumentException("accountId is required", nameof(accountId));
        }

        var id = NormalizeAccountId(accountId);
        var accountData = LoadAccount(id);

        return new ResolvedWeixinAccount
        {
            AccountId = id,
            BaseUrl = accountData?.BaseUrl?.Trim() ?? WeixinApiClient.DefaultBaseUrl,
            CdnBaseUrl = WeixinApiClient.DefaultCdnBaseUrl,
            Token = accountData?.Token?.Trim(),
            Enabled = true,
            Configured = !string.IsNullOrWhiteSpace(accountData?.Token)
        };
    }
}
